### PR TITLE
[FLINK-32349][table] Support atomic for CREATE TABLE AS SELECT(CTAS) statement

### PR DESCRIPTION
**Pull request by Tartarus0zm**
## What is the purpose of the change

Support atomic for CREATE TABLE AS SELECT(CTAS) statement

## Brief change log

Introduce StagedTable interface that support atomic operations.

## Verifying this change

StagedTableITCase#testStagedTableWithAtomicCtas

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/StagedTable.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+ *
+ * <p>CtasJobStatusHook#onCreated will call the begin method of StagedTable;
+ * CtasJobStatusHook#onFinished will call the commit method of StagedTable;
+ * CtasJobStatusHook#onFailed and CtasJobStatusHook#onCanceled will call the abort method of
+ * StagedTable;
+ */
+@PublicEvolving
+public interface StagedTable extends Serializable {
+
+    /**
```
[the remark this answers is not in the document]
Comment (luoyuxia) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/StagedTable.java:39]: No need to refer `relational database`.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/StagedTable.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+package org.apache.flink.table.catalog;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.core.execution.JobStatusHook;
+
+import java.io.Serializable;
+
+/**
+ * A {@link StagedTable} for atomic semantics using a two-phase commit protocol, combined with
+ * {@link JobStatusHook} for atomic CTAS. {@link StagedTable} will be a member variable of
```
[the remark this answers is not in the document]
Comment (luoyuxia) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/StagedTable.java:28]: Also, please elaborate the interface in the java doc as it's the core interface such as what the interface is for and who should return it And when the interface is considered, which method will be called in which condition.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/StagedTableITCase.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import org.junit.rules.TemporaryFolder;
+
+import java.io.File;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Tests staged table in a table environment. */
+public class StagedTableITCase extends BatchTestBase {
```
Comment (luoyuxia) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/StagedTableITCase.java:39]: Please test it in both batch and stream mode.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/StagedTableITCase.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+        TestCollectionTableFactory.reset();
+        TestCollectionTableFactory.initData(sourceData);
+
+        String sourceDDL = "create table t1(a int, b varchar) with ('connector' = 'COLLECTION')";
+        tEnv().executeSql(sourceDDL);
+    }
+
+    @Test
+    public void testStagedTableWithAtomicCtas() throws Exception {
```
Comment (luoyuxia) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/StagedTableITCase.java:57]: How can you make sure the data is invisible before commiting  in this test?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/factories/TestSupportsStagingTableFactory.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+            FileUtils.writeFileUtf8(
+                    new File(dataDir, "data"), value.getInt(0) + "," + value.getString(1));
+        }
+    }
+
+    /** A StagedTable for test. */
+    private static class TestStagedTable implements StagedTable {
+
+        @Override
+        public void begin() {}
```
Comment (luoyuxia) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/factories/TestSupportsStagingTableFactory.java:163]: If we do nothing in these method, we can never know these method will be really called when job begin, finish, and abort.

**Review by luoyuxia — COMMENTED** [review]
@Tartarus0zm Thanks for this pr. I left some comment.
Most is about java doc of these interfaces.
Also, the test need to be improved.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/factories/TableFactoryUtil.java [hunk 6]
```diff
@@ -164,4 +171,54 @@ public static boolean isLegacyConnectorOptions(
             }
         }
     }
+
+    /**
+     * Create a DynamicTableSink for ResolvedCatalogTable using table factory associated with the
+     * catalog.
+     */
+    public static Optional<DynamicTableSink> getDynamicTableSink(
```
[the remark this answers is not in the document]
Comment (Tartarus0zm) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/factories/TableFactoryUtil.java:179]: ok

**Comment by Tartarus0zm** [comment]
@luoyuxia  hello, I have fix your comment, PTAL

**Review by luoyuxia — COMMENTED** [review]
@Tartarus0zm Review done. Thanks for updating. PTAL

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/sink/abilities/SupportsStaging.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+import org.apache.flink.core.execution.JobStatusHook;
+import org.apache.flink.table.catalog.StagedTable;
+import org.apache.flink.table.connector.sink.DynamicTableSink;
+
+/**
+ * Interface for {@link DynamicTableSink}s that support CTAS(CREATE TABLE AS SELECT) statement
+ * atomic semantic using a two-phase commit protocol. The table sink is responsible for telling
+ * planner how to implement atomicity semantics via {@link StagedTable}.
+ *
+ * <p>If the user turns on {@link TableConfigOptions#TABLE_CTAS_ATOMICITY_ENABLED}, and {@link
```
Comment (luoyuxia) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/sink/abilities/SupportsStaging.java:31]: suggestion:

If the user turns on {@link TableConfigOptions#TABLE_CTAS_ATOMICITY_ENABLED}, and the {@link DynamicTableSink} implements {@link SupportsStaging}, the planner will call method {@link #applyStaging(StagingContext)} to get the {@link StagedTable} returned by the sink, then {@link StagedTable} will be used by Flink to implement a two-phase commit with the actual implementation of {@link StagedTable}.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/StagedTableITCase.java [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+import java.io.File;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatCode;
+
+/** Tests staged table in batch mode. */
+public class StagedTableITCase extends BatchTestBase {
```
Comment (luoyuxia) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/StagedTableITCase.java:44]:
```suggestion
public class AtomicCtasITCase extends BatchTestBase {
```
?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/StagedTableITCase.java [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+    @AfterEach
+    void close() {
+        if (tmpDataFolder != null) {
+            tmpDataFolder.delete();
+        }
+    }
+
+    @Test
+    void testStagedTableWithAtomicCtas() throws Exception {
```
Comment (luoyuxia) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/StagedTableITCase.java:71]:
```suggestion
    void testAtomicCtas() throws Exception {
```
It's not for test staged table, it's for test atomic ctas.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/StagedTableITCase.java [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+        String sourceDDL = "create table t1(a int, b varchar) with ('connector' = 'COLLECTION')";
+        tEnv().executeSql(sourceDDL);
+        tmpDataFolder = TempDirUtils.newFolder(temporaryFolder);
+    }
+
+    @AfterEach
+    void close() {
+        if (tmpDataFolder != null) {
+            tmpDataFolder.delete();
```
Comment (luoyuxia) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/StagedTableITCase.java:66]: Why delete it. Can't you provide a temPath for each method?
like:
testStagedTableWithAtomicCtas(@TempDir Path temporaryFolder)

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/sql/StagedTableITCase.java [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+import java.io.File;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatCode;
+
+/** Tests staged table in stream mode. */
+public class StagedTableITCase extends StreamingTestBase {
```
Comment (luoyuxia) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/sql/StagedTableITCase.java:44]: dito

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/StagedTableITCase.java [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+        String dataDir = tmpDataFolder.getAbsolutePath();
+        tEnv().executeSql(
+                        "create table ctas_batch_table with ('connector' = 'test-staging', 'data-dir' = '"
+                                + dataDir
+                                + "') as select * from t1")
+                .await();
+        assertThat(tEnv().listTables()).contains("ctas_batch_table");
+        // Not using StagedTable, so need to read the hidden file
+        File file = new File(dataDir, "_data");
+        assertThat(file).exists();
```
Comment (luoyuxia) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/StagedTableITCase.java:142]: Please extract 
```
assertThat(file).exists();
assertThat(file).isFile();
assertThat(FileUtils.readFileUtf8(file)).isEqualTo("1,ZM");
```
to a common method

**Review by luoyuxia — COMMENTED** [review]
@Tartarus0zm Thanks for updating. I left some comments agian as well as some typo from Grammarly. Should be ready to be merged in next few iterations.

File: docs/layouts/shortcodes/generated/table_config_configuration.html [hunk 13]
```diff
@@ -20,6 +20,12 @@
             <td>String</td>
             <td>The name of the default database in the initial catalog to be created when instantiating TableEnvironment.</td>
         </tr>
+        <tr>
+            <td><h5>table.ctas.atomicity-enabled</h5><br> <span class="label label-primary">Batch</span> <span class="label label-primary">Streaming</span></td>
+            <td style="word-wrap: break-word;">false</td>
+            <td>Boolean</td>
+            <td>Specifies if the CREATE TABLE AS SELECT statement is executed atomically. By default, the statement is non-atomic. The target table is created in client side, and it will not be dropped even though the job fails or is cancelled. If set this option to true and DynamicTableSink implements the SupportsStaging interface, the statement is expected to be executed atomically, the behavior of which depends on the actual DynamicTableSink.</td>
```
[the remark this answers is not in the document]
Comment (Tartarus0zm) [reply docs/layouts/shortcodes/generated/table_config_configuration.html:27]: IDEA boost and the previous comma can not be removed, the rest has been fixed

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java [hunk 14]
```diff
@@ -816,7 +825,56 @@ public TableResultInternal executeInternal(List<ModifyOperation> operations) {
...
+                    ExecutableOperationUtils.createDynamicTableSink(
+                            catalog,
+                            () -> moduleManager.getFactory((Module::getTableSinkFactory)),
+                            tableIdentifier,
+                            catalogTable,
+                            Collections.emptyMap(),
+                            tableConfig,
+                            resourceManager.getUserClassLoader(),
+                            createTableOperation.isTemporary());
+            if (dynamicTableSink instanceof SupportsStaging
```
[the remark this answers is not in the document]
Comment (Tartarus0zm) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java:856]: good catch

**Comment by Tartarus0zm** [comment]
@luoyuxia  hello, I have fix your comments, PTAL

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/sink/abilities/SupportsStaging.java [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+import org.apache.flink.table.connector.sink.DynamicTableSink;
+
+/**
+ * Interface for {@link DynamicTableSink}s that support atomic semantic for CTAS(CREATE TABLE AS
+ * SELECT) statement using a two-phase commit protocol. The table sink is responsible for returning
+ * a {@link StagedTable} to tell the Flink how to implement the atomicity semantics.
+ *
+ * <p>If the user turns on {@link TableConfigOptions#TABLE_CTAS_ATOMICITY_ENABLED}, and the {@link
+ * DynamicTableSink} implements {@link SupportsStaging}, the planner will call method {@link
+ * #applyStaging(StagingContext)} to get the {@link StagedTable} returned by the sink, then {@link
```
Comment (luoyuxia) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/sink/abilities/SupportsStaging.java:32]:
```suggestion
 * #applyStaging(StagingContext)} to get the {@link StagedTable} returned by the sink, then the {@link
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/sink/abilities/SupportsStaging.java [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+ * StagedTable} will be used by Flink to implement a two-phase commit with the actual implementation
+ * of {@link StagedTable}.
```
Comment (luoyuxia) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/sink/abilities/SupportsStaging.java:34]:
```suggestion
 * of the {@link StagedTable}.
```

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/AtomicCtasITCase.java [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatCode;
+
+/** Tests atomic ctas in batch mode. */
+public class AtomicCtasITCase extends BatchTestBase {
+
+    @BeforeEach
+    void setup() throws Exception {
+        super.before();
+        List<Row> sourceData = Arrays.asList(Row.of(1, "ZM"));
```
Comment (luoyuxia) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/AtomicCtasITCase.java:48]: nit:
List<Row> sourceData = Collections.singletonList(Row.of(1, "ZM"));

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/sql/AtomicCtasITCase.java [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatCode;
+
+/** Tests atomic ctas in stream mode. */
+public class AtomicCtasITCase extends StreamingTestBase {
+
+    @BeforeEach
+    void setup() throws Exception {
+        super.before();
+        List<Row> sourceData = Arrays.asList(Row.of(1, "ZM"));
```
Comment (luoyuxia) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/sql/AtomicCtasITCase.java:48]:
```suggestion
        List<Row> sourceData = Collections.singletonList(Row.of(1, "ZM"));
```

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/AtomicCtasITCase.java [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatCode;
+
+/** Tests atomic ctas in batch mode. */
+public class AtomicCtasITCase extends BatchTestBase {
```
Comment (luoyuxia) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/AtomicCtasITCase.java:43]: Also, Just found the code in stream/batch AtomicCtasITCase is almost same.
Would it make sense that we extract the code into a common abstract class like `AtomicCtasITCaseBase` just like `JoinReorderITCaseBase`? And then we will only need to rewrite `getTableEnvironment` to provided a env with stream/batch.

**Review by luoyuxia — COMMENTED** [review]
@Tartarus0zm Thanks for updating. I left minor comments. Should be good in next iteration.

**Comment by luoyuxia** [comment]
Btw, please also remember to rebase master when you fix my comments.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/StagedSinkModifyOperation.java [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+/**
+ * DML operation that tells to write to a sink which implements {@link SupportsStaging}. Currently.
+ * this operation is only for CTAS(CREATE TABLE AS SELECT) statement.
+ *
+ * <p>StagedSinkModifyOperation is an extension of SinkModifyOperation in the atomic CTAS scenario.
+ * Whiling checking whether the corresponding sink support atomic CTAS or not, we will need to get
+ * DynamicTableSink firstly and check whether it implements {@link SupportsStaging} and then call
+ * the method {@link SupportsStaging#applyStaging}. We maintain the DynamicTableSink in this
+ * operation so that we can reuse this DynamicTableSink instead of creating a new DynamicTableSink
+ * during translate the operation again which is error-prone.
```
Comment (luoyuxia) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/StagedSinkModifyOperation.java:39]: nit
```suggestion
 * during translating the operation again which is error-prone.
```

**Comment by Tartarus0zm** [comment]
@luoyuxia  hi, the test is pass, PTAL

**Review by luoyuxia — APPROVED** [review]
@Tartarus0zm Thanks for updating. LGTM.

**Comment by luoyuxia** [comment]
Please don't forget to create a jira to add a doc for the atomic CTAS statement and put the jira and FLINK-32349 under a umbrella jira.

**Comment by Tartarus0zm** [comment]
> Please don't forget to create a jira to add a doc for the atomic CTAS statement and put the jira and [FLINK-32349](https://issues.apache.org/jira/browse/FLINK-32349) under a umbrella jira.

@luoyuxia  thanks for your remind

**Pull request merged**